### PR TITLE
add validation for verified queries; add admin-app

### PR DESCRIPTION
**Pull request by sfc-gh-rehuang**
1. add table utilities;
2. add VQR validations
3. add admin app as well as update instructions for using the admin app

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -59,6 +63,18 @@ You may generate a semantic model for a given list of fully qualified tables fol
 
 All generated semantic models by default are saved either under `semantic_model_generator/output_models` if running from the root of this project or the current directory you're in.
 
+### Generation and validation using admin app
```
Comment (sfc-gh-jhilgart) [inline README.md:66]: add this under workflow section?

https://github.com/Snowflake-Labs/semantic-model-generator/pull/51/files/780c0d778644f23224350017e42ca869a5ae91c9..2f0a8cad4a512266c1d27b46497d3b8aae69ab86#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R5

File: semantic_model_generator/admin_app/app_utils.py [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+from semantic_model_generator.data_processing.proto_utils import (
+    proto_to_yaml,
+    yaml_to_semantic_model,
+)
+from semantic_model_generator.generate_model import raw_schema_to_semantic_context
+from semantic_model_generator.protos import semantic_model_pb2
+from semantic_model_generator.protos.semantic_model_pb2 import Dimension, Table
+
+
+SNOWFLAKE_ACCOUNT = os.environ["SNOWFLAKE_ACCOUNT_LOCATOR"]
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/admin_app/app_utils.py:17]: throw error if not defined?

File: semantic_model_generator/admin_app/app_utils.py [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+            model_value = getattr(st.session_state.semantic_model, field.name)
+            last_validated_value = getattr(
+                st.session_state.last_validated_model, field.name
+            )
+            if model_value != last_validated_value:
+                return True
+    return False
+
+
+# TODO:
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/admin_app/app_utils.py:44]: can you add a REAMDE under `admin_app` with this information + a brief description of the app

File: semantic_model_generator/admin_app/app_utils.py [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+    tdim.expr = st.text_input("SQL Expression", key=f"{table.name}-add-tdim-expr")
+    tdim.description = st.text_area(
+        "Description", key=f"{table.name}-add-tdim-description"
+    )
+    synonyms_df = st.data_editor(
+        pd.DataFrame(list(tdim.synonyms), columns=["Synonyms"]),
+        num_rows="dynamic",
+        key=f"{table.name}-add-tdim-synonyms",
+    )
+    del tdim.synonyms[:]
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/admin_app/app_utils.py:364]: do you need these del statements?

File: semantic_model_generator/admin_app/app_utils.py [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+    """
+    Renders a dialog box to add a new logical table.
+    """
+    table = Table()
+    table.name = st.text_input("Table Name")
+    for t in st.session_state.semantic_model.tables:
+        if t.name == table.name:
+            st.error(f"Table called '{table.name}' already exists")
+
+    table.base_table.database = st.text_input("Physical Database")
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/admin_app/app_utils.py:557]: let's use constants for `Physical Database` physical schema ..etc ?

File: semantic_model_generator/admin_app/pages/chat.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    expand_all_logical_tables_as_ctes,
+    logical_table_name,
+    remove_ltable_cte,
+)
+from semantic_model_generator.protos import semantic_model_pb2
+from semantic_model_generator.snowflake_utils.snowflake_connector import (
+    SnowflakeConnector,
+)
+
+HOST = os.environ["SNOWFLAKE_HOST"]
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/admin_app/pages/chat.py:28]: should we have a util that validates env vars required for the admin app?

File: semantic_model_generator/admin_app/pages/chat.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+) -> None:
+    """Displays a content item for a message. For generated SQL, allow user to add to verified queries directly or edit then add."""
+    message_index = message_index or len(st.session_state.messages)
+    sql = ""
+    question = ""
+    for item in content:
+        if item["type"] == "text":
+            if question == "" and "__" in item["text"]:
+                question = item["text"].split("__")[1]
+            if "<SUGGESTION>" in item["text"]:
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/admin_app/pages/chat.py:201]: nit: add pointer to where this constant is defined

File: README.md [hunk 8]
```diff
@@ -59,6 +63,18 @@ You may generate a semantic model for a given list of fully qualified tables fol
...
+
+1. Ensure you have installed the Python package. Note, the version below should be the latest version under the `dist/` directory.
+```bash
+pip install dist/semantic_model_generator-0.1.20-py3-none-any.whl
```
Comment (sfc-gh-jhilgart) [inline README.md:70]: can you update to 

`pip install dist/*.whl`

can you also update https://github.com/Snowflake-Labs/semantic-model-generator/pull/51/files/780c0d778644f23224350017e42ca869a5ae91c9..2f0a8cad4a512266c1d27b46497d3b8aae69ab86#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R82 to the same

**Review by sfc-gh-jhilgart — COMMENTED** [review]
Some comments - feel free to push back on ones you don't want to do right now

File: semantic_model_generator/admin_app/app_utils.py (hunk 2 above)
Comment (sfc-gh-rehuang) [reply semantic_model_generator/admin_app/app_utils.py]: the app will by default throws error messages on this so i think we should be good

File: semantic_model_generator/admin_app/pages/chat.py (hunk 6 above)
Comment (sfc-gh-rehuang) [reply semantic_model_generator/admin_app/pages/chat.py]: nope, we should probably add one - gonna add to TODO

File: README.md (hunk 8 above)
Comment (sfc-gh-rehuang) [reply README.md]: done!

File: semantic_model_generator/admin_app/app_utils.py (hunk 3 above)
Comment (sfc-gh-rehuang) [reply semantic_model_generator/admin_app/app_utils.py]: I added a block of instructions under the about.py st.markdown(), Can you take another look?

File: semantic_model_generator/admin_app/pages/chat.py (hunk 7 above)
Comment (sfc-gh-rehuang) [reply semantic_model_generator/admin_app/pages/chat.py]: added a comment

File: semantic_model_generator/admin_app/app_utils.py (hunk 4 above)
Comment (sfc-gh-rehuang) [reply semantic_model_generator/admin_app/app_utils.py]: I think it's needed. Also I am not gonna change the app here directly (copying from our cortex codebase); we can iterate on the app offline

File: README.md [hunk 9]
```diff
@@ -51,19 +55,35 @@ os.environ['SNOWFLAKE_ROLE'] = '<your-snowflake-role>'
...
 os.environ['SNOWFLAKE_PASSWORD'] = '<your-snowflake-password>'
+os.environ['SNOWFLAKE_ACCOUNT_LOCATOR']='<your-snowflake-account-locator>'
 os.environ['SNOWFLAKE_HOST'] = '<your-snowflake-host>'
 ```
+
+## admin-app
+
+### Generation and validation using admin app
+
+1. Ensure you have installed the Python package. Note, the version below should be the latest version under the `dist/` directory.
```
Comment (sfc-gh-jhilgart) [inline README.md:66]: remove this 

 Note, the version below should be the latest version under the `dist/` directory.

File: README.md [hunk 10]
```diff
@@ -51,19 +55,35 @@ os.environ['SNOWFLAKE_ROLE'] = '<your-snowflake-role>'
...
+```bash
+pip install dist/*.whl
+```
+2. Start the admin streamlit app on your local.
+```bash
+make admin_app
```
Comment (sfc-gh-jhilgart) [inline README.md:72]: maybe `make run_admin_app`

File: semantic_model_generator/admin_app/app_utils.py (hunk 3 above)
Comment (sfc-gh-jhilgart) [reply semantic_model_generator/admin_app/app_utils.py]: I still think readme is a more standard place to look

File: admin_app/pages/about.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import streamlit as st
```
Comment (sfc-gh-jhilgart) [inline admin_app/pages/about.py:1]: thanks for adding! an you also add admin_app/README with instructions on how to iterate / update/ how the pages work (more developer focused)

File: mypy.ini [hunk 12]
```diff
@@ -7,7 +7,7 @@ disallow_untyped_defs = True
 warn_unused_ignores = False
 disallow_any_generics = True
 
-exclude = pyvenv|(_test\.py|test_.*\.py)|_pb2\.py|_pb2\.pyi
+exclude = pyvenv|(_test\.py|test_.*\.py)|_pb2\.py|_pb2\.pyi|admin_app/
```
Comment (sfc-gh-jhilgart) [inline mypy.ini:10]: We should have mypy for streamlit apps I think

File: semantic_model_generator/data_processing/cte_utils.py [hunk 13]
```diff
@@ -0,0 +1,227 @@
+from typing import List
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/data_processing/cte_utils.py:1]: please add tests for this file (or, ideally, break into a new PR)

File: admin_app/pages/about.py (hunk 11 above)
Comment (sfc-gh-rehuang) [reply admin_app/pages/about.py:1]: sure added

File: semantic_model_generator/data_processing/cte_utils.py (hunk 13 above)
Comment (sfc-gh-rehuang) [reply semantic_model_generator/data_processing/cte_utils.py:1]: those utils are needed for the streamlit code to work. So put it here for now

File: mypy.ini (hunk 12 above)
Comment (sfc-gh-rehuang) [reply mypy.ini:10]: so I moved our streamlit code from cortex repo which's under 3.10 and here we use python 3.8 which i need to update a lot of places for the lint to work... if you don't mind i'll push lint to later. And i think for larger changes, make sense to do it in cortex repo first and then move to here.

**Review by sfc-gh-jhilgart — APPROVED** [review]
lgtm - please add linting as a fast follow

**Pull request merged**